doc/api/async_hooks.md

```diff
@@ -0,0 +1,677 @@
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running the constructor, and must be explicitly run to begin
+// executing callbacks.
+asyncHook.enable();
+
+// Disable listening for new asynchronous events.
+asyncHook.disable();
+
+//
+// The following are the callbacks that can be passed to createHook().
+//
+
+// init() is called during object construction. The resource may not have
+// completed construction when this callback runs, therefore all fields of the
+// resource referenced by "id" may not have been populated.
+function init(id, type, triggerId, resource) { }
+
+// before() is called just before the resource's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWrap), and should be called exactly 1
+// time for requests (e.g. FSReqWrap).
+function before(id) { }
+
+// after() is called just after the resource's callback has finished.
+// This hook will not be called if an uncaught exception occurred.
+function after(id) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPParser where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after after() has
+// completed.
+function destroy(id) { }
+```
+
+#### `async_hooks.createHook(callbacks)`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+* `callbacks` {Object} the callbacks to register
+
+Registers functions to be called for different lifetime events of each async
+operation.
+The callbacks `init()`/`before()`/`after()`/`destroy()` are registered via an
+`AsyncHooks` instance and fire during the respective asynchronous events in the
+lifetime of the event loop. The focal point of these calls centers around the
+lifetime of the `AsyncWrap` C++ class. These callbacks will also be called to
+emulate the lifetime of handles and requests that do not fit this model. For
+example, `HTTPParser` instances are recycled to improve performance. Therefore the
+`destroy()` callback is called manually after a connection is done using
+it, just before it is placed back into the unused resource pool.
+
+All callbacks are optional. So, for example, if only resource cleanup needs to
+be tracked then only the `destroy()` callback needs to be passed. The
+specifics of all functions that can be passed to `callbacks` is in the section
+`Hook Callbacks`.
+
+##### Error Handling
+
+If any `AsyncHook` callbacks throw, the application will
+print the stack trace and exit. The exit path does follow that of any uncaught
+exception. However `'exit'` callbacks will still fire unless the application
+is run with `--abort-on-uncaught-exception`, in which case a stack trace will
+be printed and the application exits, leaving a core file.
+
+The reason for this error handling behavior is that these callbacks are running
+at potentially volatile points in an object's lifetime. For example, during
```

Comment:
`object's lifetime. For example, during` → `object's lifetime, for example during`